cli/urfave/mod.go

```diff
@@ -0,0 +1,165 @@
+// Package urfave provides a cli builder implementation based on the urfave/cli
+// library.
+package urfave
+
+import (
+	"fmt"
+
+	ucli "github.com/urfave/cli/v2"
+	"go.dedis.ch/dela/cli"
+)
+
+// Builder implements a cli builder based on urfave/cli
+//
+// - implements cli.Builder
+type Builder struct {
+	commands map[string]*cmdBuilder
+	name     string
+	action   cli.Action
+	flags    []cli.Flag
+}
+
+// NewBuilder returns a new initialized builder. Action allows one to define a
+// primary action, but can be nil if we only needs to define commands. Flags
+// provides the global flags available from all the commands/subcommands.
+func NewBuilder(name string, action cli.Action, flags ...cli.Flag) cli.Builder {
+	return &Builder{
+		name:     name,
+		commands: make(map[string]*cmdBuilder),
+		action:   action,
+		flags:    flags,
+	}
+}
+
+// Build implements cli.builder.
+func (b Builder) Build() cli.Application {
+	app := &ucli.App{
+		Name:     b.name,
+		Commands: buildCommand(b.commands),
+		Action:   makeAction(b.action),
+		Flags:    buildFlags(b.flags),
+	}
+
+	app.Setup()
+
+	return app
+}
+
+// SetCommand implements cli.Builder.
+func (b *Builder) SetCommand(name string) cli.CommandBuilder {
+	cmd := &cmdBuilder{
+		subcommands: make(map[string]*cmdBuilder),
+	}
+	b.commands[name] = cmd
+
+	return cmd
+}
+
+// commandBuilder is the struct provided to build commands.
+//
+// - implements cli.CommandBuilder
+type cmdBuilder struct {
+	description string
+	action      cli.Action
+	flags       []ucli.Flag
+	subcommands map[string]*cmdBuilder
+}
+
+// SetDescription implements cli.CommandBuilder.
+func (b *cmdBuilder) SetDescription(value string) {
+	b.description = value
+}
+
+// SetFlags implements cli.CommandBuilder.
+func (b *cmdBuilder) SetFlags(flags ...cli.Flag) {
+	b.flags = buildFlags(flags)
+}
+
+// SetAction implements cli.CommandBuilder.
+func (b *cmdBuilder) SetAction(action cli.Action) {
+	b.action = action
+}
+
+// SetSubCommand implements cli.CommandBuilder.
+func (b *cmdBuilder) SetSubCommand(name string) cli.CommandBuilder {
+	builder := &cmdBuilder{
+		subcommands: make(map[string]*cmdBuilder),
+	}
+	b.subcommands[name] = builder
```

Comment:
shouldn't we do error-checking to verify that `b.subcommands[name]` hasn't already been defined?